### PR TITLE
WT-10383 Make asserts and function calls consistent with runtime assertion checks

### PR DESCRIPTION
**Pull request by clarissecheah**

### CODE REVIEW

File: src/txn/txn.c [hunk 1]
```diff
@@ -1453,7 +1454,8 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
              */
             if (!commit && resolve_case == RESOLVE_PREPARE_ON_DISK &&
               head_upd->type == WT_UPDATE_STANDARD && F_ISSET(head_upd, WT_UPDATE_RESTORED_FROM_HS))
-                WT_ASSERT(session, head_upd->next == NULL);
+                WT_ASSERT_ALWAYS(session, head_upd->next == NULL,
+                  "Failed to correctly restore update from history store");
```
Comment (ajmorton) [inline src/txn/txn.c:1458]: I'm not sure what this assert is protecting against.. It's either that we've restored multiple updates from the history store or we've placed the history update at the wrong location in the chain

File: src/txn/txn_log.c [hunk 2]
```diff
@@ -39,14 +39,17 @@ __txn_op_log_row_key_check(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt)
...
         key.data = WT_INSERT_KEY(cbt->ins);
         key.size = WT_INSERT_KEY_SIZE(cbt->ins);
     }
 
-    WT_ASSERT(session, __wt_compare(session, CUR2BT(cbt)->collator, &key, &cursor->key, &cmp) == 0);
-    WT_ASSERT(session, cmp == 0);
+    WT_ASSERT_ALWAYS(session,
+      __wt_compare(session, CUR2BT(cbt)->collator, &key, &cursor->key, &cmp) == 0,
+      "Comparison of row store logging key and cursor key failed");
+    WT_ASSERT_ALWAYS(session, cmp == 0, "Row store logging key and cursor key mismatch");
```
Comment (ajmorton) [inline src/txn/txn_log.c:52]: "Cursor is referencing the wrong key"

**Review by ajmorton — COMMENTED** [review]
Thanks!
The  `__cursor_key_order_check_col`, `__cursor_key_order_check_row`,  `__wt_cursor_key_order_init`, `__wt_cursor_key_order_reset`, `__wt_btcur_open`,  `__wt_btcur_close` calls look good now they're all hidden behind `DIAGNOSTIC_ASSERTS_ENABLED`.

For the assert messages I've provided a few alternate suggestions that are hopefully more specific, but let's get a second pair of eyes on these to confirm.

File: src/btree/bt_curnext.c [hunk 3]
```diff
@@ -980,7 +980,8 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
                 F_CLR(cursor, WT_CURSTD_BOUND_LOWER_INCLUSIVE);
                 ret = __wt_compare_bounds(
                   session, cursor, &cbt->iface.key, cbt->recno, false, &key_out_of_bounds);
-                WT_ASSERT(session, ret == 0 && !key_out_of_bounds);
+                WT_ASSERT_ALWAYS(session, ret == 0 && !key_out_of_bounds,
+                  "Failed to verify cursor position after walking. Cursor is out of bounds");
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/btree/bt_curnext.c:984]: Perhaps: "Bounded cursor logic resulted in the cursor being an illegal state." This assertion was intentionally in a #HAVE_DIAGNOSTIC block because it validates prior logic but is somewhat expensive. If this assertion fails explaining why that is bad is hard so I think a specific message won't be very valuable.

File: src/btree/bt_curprev.c [hunk 4]
```diff
@@ -912,7 +912,8 @@ __wt_btcur_prev(WT_CURSOR_BTREE *cbt, bool truncating)
                 F_CLR(cursor, WT_CURSTD_BOUND_UPPER_INCLUSIVE);
                 ret = __wt_compare_bounds(
                   session, cursor, &cbt->iface.key, cbt->recno, true, &key_out_of_bounds);
-                WT_ASSERT(session, ret == 0 && !key_out_of_bounds);
+                WT_ASSERT_ALWAYS(session, ret == 0 && !key_out_of_bounds,
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/btree/bt_curprev.c:915]: See my previous comment.

File: src/btree/bt_cursor.c [hunk 5]
```diff
@@ -847,10 +847,9 @@ __wt_btcur_search(WT_CURSOR_BTREE *cbt)
     if (session->format_private != NULL && (ret == 0 || ret == WT_NOTFOUND))
         session->format_private(ret, session->format_private_arg);
 
-#ifdef HAVE_DIAGNOSTIC
-    if (ret == 0)
-        WT_ERR(__wt_cursor_key_order_init(cbt));
-#endif
+    if (DIAGNOSTIC_ASSERTS_ENABLED(session))
```
Comment (luke-pearson) [inline src/btree/bt_cursor.c:850]: Is there a reason we chose to not merge these if's? i.e.
```
    if (DIAGNOSTIC_ASSERTS_ENABLED(session) && ret == 0)
            WT_ERR(__wt_cursor_key_order_init(cbt));
```
Similarly for __btcur_search_neighboring.

File: src/btree/bt_split.c [hunk 6]
```diff
@@ -69,9 +69,10 @@ __split_verify_intl_key_order(WT_SESSION_IMPL *session, WT_PAGE *page)
     case WT_PAGE_COL_INT:
         recno = 0; /* Less than any valid record number. */
         WT_INTL_FOREACH_BEGIN (session, page, ref) {
-            WT_ASSERT(session, ref->home == page);
-
-            WT_ASSERT(session, ref->ref_recno > recno);
+            WT_ASSERT_ALWAYS(
+              session, ref->home == page, "Failed to verify page reference after splitting");
```
Comment (luke-pearson) [inline src/btree/bt_split.c:73]: I think we can have more detail here: "Internal page in illegal state, child ref is referencing an incorrect page"

File: src/btree/bt_split.c [hunk 7]
```diff
@@ -69,9 +69,10 @@ __split_verify_intl_key_order(WT_SESSION_IMPL *session, WT_PAGE *page)
...
+            WT_ASSERT_ALWAYS(
+              session, ref->ref_recno > recno, "Failed to verify valid recno after splitting");
```
Comment (luke-pearson) [inline src/btree/bt_split.c:75]: Something like: "Incorrectly sorted refs detected in a parent index."

File: src/btree/bt_split.c [hunk 8]
```diff
@@ -84,16 +85,19 @@ __split_verify_intl_key_order(WT_SESSION_IMPL *session, WT_PAGE *page)
 
         slot = 0;
         WT_INTL_FOREACH_BEGIN (session, page, ref) {
-            WT_ASSERT(session, ref->home == page);
+            WT_ASSERT_ALWAYS(
+              session, ref->home == page, "Failed to verify page reference after splitting");
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/btree/bt_split.c:89]: This is doing the same this as the COL_INT check, effectively when a page splits it's child refs are moved between pages. This check ensures that the ref->home is correctly referring to the parent page, and not the previous parent.

File: src/btree/bt_split.c [hunk 9]
```diff
@@ -84,16 +85,19 @@ __split_verify_intl_key_order(WT_SESSION_IMPL *session, WT_PAGE *page)
...
             /*
              * Don't compare the first slot with any other slot, it's ignored on row-store internal
              * pages.
              */
             __wt_ref_key(page, ref, &next->data, &next->size);
             if (++slot > 2) {
-                WT_ASSERT(session, __wt_compare(session, btree->collator, last, next, &cmp) == 0);
-                WT_ASSERT(session, cmp < 0);
+                WT_ASSERT_ALWAYS(session,
+                  __wt_compare(session, btree->collator, last, next, &cmp) == 0,
```
Comment (luke-pearson) [inline src/btree/bt_split.c:98]: It is unusual but `__wt_compare` can return an error code, so what this is checking is that the call to the comparison function returns successfully as the subsequent check depends on it. A better message may be: "Key comparison failed during verification, cannot proceed."

File: src/cursor/cur_hs.c [hunk 10]
```diff
@@ -998,14 +997,17 @@ __curhs_insert(WT_CURSOR *cursor)
     } while ((ret = __wt_hs_modify(cbt, hs_upd)) == WT_RESTART);
     WT_ERR(ret);
 
-    /* We no longer own the update memory, the page does; don't free it under any circumstances. */
+    /*
+     * We no longer own the update memory, the page does; don't free it under any circumstances.
```
Comment (luke-pearson) [inline src/cursor/cur_hs.c:1001]: Why did this get multi-lined?

Comment (clarissecheah) [reply src/cursor/cur_hs.c]: Not sure, s_all did that but I've reverted it: [4ec1860](https://github.com/wiredtiger/wiredtiger/pull/8623/commits/4ec1860657e6aeef19c7f7a20fa12817c7c791a1)

File: src/btree/bt_split.c (hunk 9 above)
Comment (clarissecheah) [reply src/btree/bt_split.c:98]: [4ec1860](https://github.com/wiredtiger/wiredtiger/pull/8623/commits/4ec1860657e6aeef19c7f7a20fa12817c7c791a1)

File: src/btree/bt_cursor.c (hunk 5 above)
Comment (clarissecheah) [reply src/btree/bt_cursor.c]: No specific reason but was an oversight. Thanks! [4ec1860](https://github.com/wiredtiger/wiredtiger/pull/8623/commits/4ec1860657e6aeef19c7f7a20fa12817c7c791a1)

- [ ]

File: src/btree/bt_curnext.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (clarissecheah) [reply src/btree/bt_curnext.c]: [4ec1860](https://github.com/wiredtiger/wiredtiger/pull/8623/commits/4ec1860657e6aeef19c7f7a20fa12817c7c791a1)

[the remark this answers is not in the document]
Comment (ajmorton) [reply src/btree/bt_curnext.c]: @luke-pearson the `#HAVE_DIAGNOSTIC` block has been replaced with a runtime `DIAG_ASSERTS_ENABLED` check so we still skip these checks unless the user sets a flag on the connection. The change from `ASSERT` to `ASSERT_ALWAYS` is so this code will actually run when we turn on extra asserts at runtime.

r.e. `explaining why that is bad is hard` would it be useful to turn this on in production? If not we can change this back to a `#HAVE_DIAGNOSTIC` block which is used in testing only.

[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/btree/bt_curnext.c]: This assertion makes sense as a runtime enableable assertion. I assume if a user was turning that on it'd likely be at our request so we could parse the message. That's why I don't think we need a lot of detail here.

File: src/btree/bt_cursor.c (hunk 5 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c]: @ajmorton you made a node that the DIAGNOSTIC_ASSERTS check pulls a pointer off the session, is it better to check the `ret` first?

File: src/btree/bt_curnext.c [hunk 11]
```diff
@@ -980,7 +980,8 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
                 F_CLR(cursor, WT_CURSTD_BOUND_LOWER_INCLUSIVE);
                 ret = __wt_compare_bounds(
                   session, cursor, &cbt->iface.key, cbt->recno, false, &key_out_of_bounds);
-                WT_ASSERT(session, ret == 0 && !key_out_of_bounds);
+                WT_ASSERT_ALWAYS(session, ret == 0 && !key_out_of_bounds,
+                  "Bounded cursor logic resulted in the cursor being in an illegal state");
```
Comment (luke-pearson) [inline src/btree/bt_curnext.c:984]: Would it be worthwhile making these assert messages differentiable? So for this one: "Bounded cursor next logic..." and for the prev one "Bounded cursor prev logic...".

File: src/btree/bt_cursor.c [hunk 12]
```diff
@@ -2260,10 +2256,10 @@ __wt_btcur_open(WT_CURSOR_BTREE *cbt)
     cbt->upd_value->type = WT_UPDATE_INVALID;
     WT_TIME_WINDOW_INIT(&cbt->upd_value->tw);
 
-#ifdef HAVE_DIAGNOSTIC
-    cbt->lastkey = &cbt->_lastkey;
-    cbt->lastrecno = WT_RECNO_OOB;
-#endif
+    if (DIAGNOSTIC_ASSERTS_ENABLED((WT_SESSION_IMPL *)cbt->iface.session)) {
```
Comment (luke-pearson) [inline src/btree/bt_cursor.c:2259]: This cast should be doable with CUR2S(cbt).

File: src/cursor/cur_hs.c [hunk 13]
```diff
@@ -998,14 +997,17 @@ __curhs_insert(WT_CURSOR *cursor)
...
+     */
     hs_tombstone = hs_upd = NULL;
 
     if (DIAGNOSTIC_ASSERTS_ENABLED(session)) {
         /* Do a search again and call next to check the key order. */
         ret = __curhs_file_cursor_search_near(session, file_cursor, &exact);
         /* We can get not found if the inserted history store record is obsolete. */
-        WT_ASSERT(session, ret == 0 || ret == WT_NOTFOUND);
+        WT_ASSERT_ALWAYS(session, ret == 0 || ret == WT_NOTFOUND,
+          "Search near validation of inserted record failed with: %d", ret);
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/cursor/cur_hs.c:1007]: How about: "Key order verification failed for history store insertion."

File: src/history/hs_rec.c [hunk 14]
```diff
@@ -164,9 +164,10 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
                          * not idempotent, having them inserted multiple times can cause invalid
                          * values to be read.
                          */
-                        WT_ASSERT(session,
+                        WT_ASSERT_ALWAYS(session,
                           type != WT_UPDATE_MODIFY &&
-                            (uint8_t)upd_type_full_diag != WT_UPDATE_MODIFY);
+                            (uint8_t)upd_type_full_diag != WT_UPDATE_MODIFY,
+                          "Inserted invalid duplicate record of type modify update");
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/history/hs_rec.c:170]: Maybe: "Duplicate modifies inserted into the history store result in invalid reads."

[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/history/hs_rec.c:170]: Its hard to gauge what level of detail is useful here, I think the original comment is pretty good as well.

File: src/btree/bt_cursor.c (hunk 5 above)
Comment (ajmorton) [reply src/btree/bt_cursor.c]: Interesting point..
The `ASSERT_ENABLED` is wrapped inside an `UNLIKELY` under the hood so branch prediction will always assume it's false. My gut feel is that the order of args won't make a big difference at runtime.

To add to that we've scaled back the use of our runtime flag from pretty much every `WT_ASSERT` in the codebase to only a subset of locations like this one, so the impacts of pointer derefs will be much less than the 10% perf drop we saw previously. This also doesn't look like a critical code path (the actual searcing will take much longer) so we can won't need to worry too much.

Finally on a stylistic note I like having `ASSERT_ENABLED` as the first arg. To me it indicates a context switch from normal code flow to diagnostic mode which is more relevant to the reader than the `ret` check. I'd even suggest keeping these two conditions as separate `if`s for readability.

Comment (luke-pearson) [reply src/btree/bt_cursor.c]: I am happy to leave it as is then I'd prefer not to split it back into two if statements. One thing worth calling out is that transforming `#ifdef HAVE_DIAGNOSTIC` into a `DIAGNOSTIC_ASSERTS_ENABLED` check is a bit odd when the what the diagnostic code was doing was setting it up to check something at a later date, or even just wrapping some expensive code. If you understand the context that by default `DIAGNOSTIC_ASSERTS_ENABLED` is true for a `DIAGNOSTIC` build then it is clear what it is doing but it requires extra context than the original `#ifdef HAVE_DIAGNOSTIC` block. Not sure what we could do here as adding a code comment for every usage of it seems excessive.

File: src/history/hs_rec.c [hunk 15]
```diff
@@ -1097,16 +1098,25 @@ __hs_delete_record(
     } else {
         if (DIAGNOSTIC_ASSERTS_ENABLED(session)) {
             __wt_hs_upd_time_window(r->hs_cursor, &hs_tw);
-            WT_ASSERT(session, hs_tw->start_txn == WT_TXN_NONE || hs_tw->start_txn == upd->txnid);
-            WT_ASSERT(session, hs_tw->start_ts == WT_TS_NONE || hs_tw->start_ts == upd->start_ts);
-            WT_ASSERT(session,
-              hs_tw->durable_start_ts == WT_TS_NONE || hs_tw->durable_start_ts == upd->durable_ts);
+            WT_ASSERT_ALWAYS(session,
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/history/hs_rec.c:1101]: This entire codepath was added in WT-9805 which was a recent change I'd suggest giving @quchenhao context and getting his opinion on what messages are best here.

File: src/log/log_slot.c [hunk 16]
```diff
@@ -24,7 +24,7 @@ __log_slot_dump(WT_SESSION_IMPL *session)
     conn = S2C(session);
     log = conn->log;
     ret = __wt_verbose_dump_log(session);
-    WT_ASSERT(session, ret == 0);
+    WT_ASSERT_ALWAYS(session, ret == 0, "Failed to dump logging information");
```
Comment (luke-pearson) [inline src/log/log_slot.c:27]: This is actually dumping information about the logging subsystem so I'd reword to: "Failed to dump log system information."

File: src/btree/bt_cursor.c (hunk 5 above)
Comment (ajmorton) [reply src/btree/bt_cursor.c]: 👍  
WT-10428 created to update the `DIAGNOSTIC_ASSERTS_ENABLED` to something better.

File: src/history/hs_rec.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/history/hs_rec.c:1101]: This code is called in the case after a prepared rollback in reconciliation. In prepared rollback, if the prepared update is evicted to disk, we will first bring back the previous update from the history store and append it to the update chain. Later in the reconciliation, we will remove that update from the history store as it is now on the update chain. Here is to assert we are removing the update with the correct transaction id and timestamp.

File: src/reconcile/rec_write.c [hunk 17]
```diff
@@ -821,28 +821,33 @@ __rec_write(WT_SESSION_IMPL *session, WT_ITEM *buf, uint8_t *addr, size_t *addr_
 
     if (DIAGNOSTIC_ASSERTS_ENABLED(session)) {
         /* Checkpoint calls are different than standard calls. */
-        WT_ASSERT(session,
+        WT_ASSERT_ALWAYS(session,
           (!checkpoint && addr != NULL && addr_sizep != NULL) ||
-            (checkpoint && addr == NULL && addr_sizep == NULL));
+            (checkpoint && addr == NULL && addr_sizep == NULL),
+          "Failed to verify checkpoint call");
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/reconcile/rec_write.c:827]: This was a bit of dive into the code but when this is called for a "checkpoint" we are writing out a root page as far as I can tell and it has special logic to create a checkpoint. This doesn't require an `addr` and associated `addr_sizep`. So this call is checking that when called for a checkpoint it is passed the correct arguments. 

Maybe we need: "Incorrect arguments passed to rec_write for a checkpoint call". It should really only fire in dev while updating a caller to this function.

File: src/txn/txn.c [hunk 18]
```diff
@@ -1440,9 +1440,10 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
              * Assert if we still have an update from the current transaction that hasn't been
              * resolved or aborted.
              */
-            WT_ASSERT(session,
+            WT_ASSERT_ALWAYS(session,
               head_upd->txnid == WT_TXN_ABORTED || head_upd->prepare_state == WT_PREPARE_RESOLVED ||
-                head_upd->txnid != txn->id);
+                head_upd->txnid != txn->id,
+              "Current transaction has updates that haven't been resolved or aborted");
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/txn/txn.c:1446]: I think this is checking the following: "Failed to resolve all updates associated with a prepared transaction."

File: src/txn/txn.c (hunk 1 above)
Comment (luke-pearson) [reply src/txn/txn.c:1458]: We're rolling back a prepared transaction as indicated by the `!commit` check, I'd say that this is ensuring that we correctly placed the update prior to the prepared transaction on the update chain. As it was restored from the history store it wouldn't make sense to have something "older" than it on the chain. 

Perhaps we can word it as follows: "Rolling back a prepare transaction resulted in an invalid update chain."

File: src/txn/txn_log.c [hunk 19]
```diff
@@ -39,14 +39,17 @@ __txn_op_log_row_key_check(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt)
         page = cbt->ref->page;
         WT_ASSERT(session, cbt->slot < page->entries);
         rip = &page->pg_row[cbt->slot];
-        WT_ASSERT(session, __wt_row_leaf_key(session, page, rip, &key, false) == 0);
+        WT_ASSERT_ALWAYS(session, __wt_row_leaf_key(session, page, rip, &key, false) == 0,
+          "Failed to set a buffer for row-store leaf page key");
```
Comment (luke-pearson) [inline src/txn/txn_log.c:43]: "Failed to instantiate a row-store key, cannot proceed with cursor key verification."

File: src/txn/txn_log.c (hunk 2 above)
Comment (luke-pearson) [reply src/txn/txn_log.c:52]: "Cursor is not referencing the expected key when logging an operation." might work too.

File: src/history/hs_rec.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (clarissecheah) [reply src/history/hs_rec.c:170]: It feels like this assert is checking that the duplicate HS records aren't modify updates since this _can_ lead to reading invalid values. Do you think the suggested assertion message makes it sound as though the invalid read has already occurred instead of verifying the dup record type?

File: src/history/hs_rec.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (clarissecheah) [reply src/history/hs_rec.c:1101]: Thanks for the explanation @quchenhao. Open question, would messages like these serve the purpose better? 

"Reconciliation attempted to remove incorrect HS update: unexpected transaction id" 
"Reconciliation attempted to remove incorrect HS update: unexpected timestamp"

File: src/history/hs_rec.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (ajmorton) [reply src/history/hs_rec.c:170]: Yeah that's a good point. Let's go with Luke's suggestion above

[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/history/hs_rec.c:170]: In that case: "Duplicate modifies inserted into the history store can result in invalid reads."

File: src/history/hs_rec.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (ajmorton) [reply src/history/hs_rec.c:1101]: Thanks @quchenhao!
How about we add a comment above the asserts explaining the context, then keep the assert messages simple. Something like:
``` 
/* 
 * If we're deleting a record that is already in the history store this implies we're rolling back a prepared transaction which 
 * means we need to pull the update from the history store back into the update chain and then delete it from the history store.
 * These checks ensure we've retrieved the correct update from the history store.
 */
```
and then the error messages can be
`Retrieved wrong update from history store: XXX mismatch`

File: src/btree/bt_curnext.c (hunk 11 above)
Comment (clarissecheah) [reply src/btree/bt_curnext.c]: Yep, offers the users more context. Done in: [9642025](https://github.com/wiredtiger/wiredtiger/pull/8623/commits/9642025e6b96c3914f80e7e39cf4758ad4637814). Thanks!

File: src/history/hs_rec.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (quchenhao) [reply src/history/hs_rec.c:1102]: SGTM.

[the remark this answers is not in the document]
Comment (clarissecheah) [reply src/history/hs_rec.c:1108]: [a002800](https://github.com/wiredtiger/wiredtiger/pull/8623/commits/a00280075838de50f7a0b4c26ed04114fb449698)

**Review by ajmorton — APPROVED** [review]
LGTM. Thanks!
